src/components/FullHierarchyTable/FullHierarchyTable.tsx

```diff
@@ -56,8 +75,88 @@ const Exp: FunctionComponent = (props: any) => {
 export const FullHierarchyTable: React.FunctionComponent<FullHierarchyTableAllProps> = (props) => {
     const bcName = props.meta.bcName
     const fields = props.meta.fields
+    const loading = props.loading
     const depthLevel = props.depth || 1
     const indentLevel = depthLevel - 1
+    const {t} = useTranslation()
+    const [userOpenedRecords, setUserOpenedRecords] = React.useState([])
+
+    React.useEffect(
+        () => {
+            if (props?.expandedRowKeys) {setUserOpenedRecords(props.expandedRowKeys)}
+        },
+        [props.expandedRowKeys]
+    )
+    const [inputPlaceholder, setInputPlaceholder] = React.useState(props?.searchPlaceholder || '')
+
+    // Apply one filter item
+    const checkFiltered = (filter: BcFilter, checkData: AssociatedItem[]) => {
+        const filteredData: AssociatedItem[] = []
+        checkData?.forEach((dataItem) => {
+            if (dataItem[filter?.fieldName]?.toString().toLowerCase()?.indexOf(filter?.value?.toString().toLowerCase()) > -1) {
+                if (filter) {
+                    dataItem._searched = true
+                }
+                filteredData.push(dataItem)
+            }
+        })
+        return filteredData
+    }
+
+    // Function check tree to find all parent
+    const findParent = (filteredData: AssociatedItem[], checkData: AssociatedItem[]) => {
+        const parentData: AssociatedItem[] = []
+        filteredData?.forEach((dataItem) => {
+            let currentLevel = dataItem.level as number
+            let currentParentId = dataItem.parentId as string
+            while (currentLevel > 1) {
+                const parentItem = props.data.find((item) => item.id === currentParentId)
+                parentData.push(parentItem)
+                currentParentId = parentItem.parentId as string
+                currentLevel--
+            }
+        })
+        return parentData
+    }
+
+    // Function check tree to find all child
+    const findChild = (filteredData: AssociatedItem[], checkData: AssociatedItem[]) => {
```

Comment:
I have literally no idea what is going on here and so as any other developer I suspect.  
1) You do not have a tree structure to traverse to begin with.
2) Method signature is a mess. BFS assumes you have a tree node and a condition to check the descendants for. Here you have two arrays, the latter is an array for a search (which is unclear until you look at the method call) and the former is god knows what: some array of uncertain nodes that is mutated inside the method. What was the input? It's even more obscure when you look at the method call and see that this mutable array was also previously mutated by another handler and shared in a spread.
3) What is the point of all this compared to a simple `props.data(child => child.parentId === parent.id)`?

Also you've completely neglected our [contributing guide](https://github.com/tesler-platform/tesler-ui/blob/master/CONTRIBUTING.md) which prescribes to put utility helpers in the end of component, precisely for the reason to avoid situations like this when the component becomes completely unreadable.

Search algorithm should be isolated, should have a clear contract and should be isolated from the setState calls that you use for expanding found nodes.
I believe following steps will fix the situation:

**Step 1**. Start with the interface, the code clearly lying about nodes being AssociatedItem:
```tsx
interface FullHierarchyDataItem extends AssociatedItem {
    parentId: string,
    level: number
}
```

**Step 2**. You should limit the filters according to the real scenario: your search highlight will not work for any type of filters and you `.toString()` calls will lead to `[object Object]` or null pointer exceptions, so it should be clear from the code that you only handle Input/TextArea and stringified numbers, with filters only set on 'contains' and 'equals' search types:
```tsx
const textFilters = React.useMemo(() => props.filters.find(filter =>
        [FilterTypes.contains, FilterTypes.equals].includes(filter.type),
    [props.filters]
)
```

**Step 3**. You should implement a condition check for one item, probably move it to a different module. Similiar to you  `checkFiltered` function, but more isolated and without data item mutations and introducing another obscure `_search` flag:
```tsx
/* JSDoc about type limitations and comparing logic (stringified numbers, lowercase strings), unit tests */
function bcFilterTextMatch(dataItem: FullHierarchyDataItem, filters: BcFilter[]) {
    if (filters?.length === 0) {
        return true
    }
    return filters?.every(filter => {
        const filterable = typeof dataItem[filter.fieldName] === 'string' || typeof dataItem[filter.fieldName] === 'number'
        const searchExpression = String(filter.value).toLowerCase()
        const value = String(dataItem[filter.fieldName]).toLowerCase()
        return filterable && value.includes(searchExpression)
    })
}
```

**Step 4**. If you have a match, that means all ancestors are matched automatically. Also, it is not clear from the code but after testing your implementation it seems that in some cases all sibling nodes for matched node also should be matched.
```tsx
/* JSDoc about how the search works and what items will be included in a result set, unit tests */
function bcFilterMatchedAncestors(dataItem: FullHierarchyDataItem, dataItems: FullHierarchyDataItem[], filters: BcFilter[]) {
    // Array of keys probably will be enough
    const result: string[] = []
    // You check the item
    if (bcFilterTextMatch(dataItem, filters)) {
        let current = dataItem
        // In case you need to keep siblings, additional logic like "only for the last level" should be here
        dataItems.filter(sibling => sibling.parentId === current.parentId).forEach(sibling => result.push(sibling.id))
        do {
            // You store the match
            result.push(current.id)
            // And traverse through ancestors to store them too
            current = dataItems.find(item => item.id === current.parentId)
        } while(current?.parentId)
    }
    return result
}
```

**Step 5**. You iterate through the data items to get the match and corresponding ancestors, then get rid of the duplicates with Set, result is memoized:
```tsx
const searchedKeys: Set<string> = React.useMemo(() => {
    const result: string[] = []
    props.data.forEach(item => {
        bcFilterMatchedAncestors(item, props.data, props.bcFilter)?.forEach(key => result.push(key))
    })
    return new Set(result)
}, [props.data, props.bcFilter])
```
You can also map keys to items in case you need them:
```tsx
const filteredData = React.useMemo(() => {
    return props.bcFilter?.length
        ? props.data.filter(item => searchedKeys.has(item.id))
        : props.data
}, [searchedKeys, props.data, props.bcFilter])
```
And that's all, now you have a filtered subset of data. Any optimization should be handled with care because if things will go that way then filtration will either be delegated to Tesler API or will require some different data structures and algorithms. In any case, it all should come after measurements and with implementations thoroughly documented and tested.

**Step 6**. Now in case you need to automatically expand all matching nodes, current implementation is too messy so I just give an overview: you take the filtered result set and than exclude the records with no children, as they should be just shown, not expand. You put it in an existing effect already handling pre-expanded rows:
```tsx
React.useEffect(
    () => {
        // Could be `preopenedRecords` instead, currently it's unclear from the code where is what
        const manualOpen = props.expandedRowKeys || []
        const expandFiltered = props.filters?.length
            // Filter out leafs without children
            ? filteredData.filter(item => filteredData.some(child => item.id === child.parentId)).map(item => item.id)
            : []
        const newExpandedKeys = new Set([ ...manualOpen, ...expandFiltered ])
        if (newExpandedKeys.size > 0) {setUserOpenedRecords(Array.from(newExpandedKeys))}
    },
    [props.expandedRowKeys, props.bcFilter, filteredData]
)
```

Reply:
Thanks for the code review and comments, they are relevant. The component was really confusing, the intention to simplify it is obvious.
I followed the steps listed and got the result. However, I have comments:

**Step 2**  `props.filters?.find`
```ts
const textFilters = React.useMemo(() => props.filters?.find(filter =>
        [FilterTypes.contains, FilterTypes.equals].includes(filter.type),
    [props.filters]
)
```

Now `props.bcFilter` replace `textFilters`


**Step 5** 
```ts
bcFilterMatchedAncestors(item, props.data, props.bcFilter)?.forEach(key => result.push(key))`
Type mismatch, explicit casting is needed:
`bcFilterMatchedAncestors(item as FullHierarchyDataItem, props.data as FullHierarchyDataItem[], textFilters)
```

The proposed data filter algorithm does not take into account the fact that both ancestors and descendants of data elements that fall under the search conditions must be present in the result set. This is an important point.

I divided the key constants of the ancestors and descendants so that they could be used separately (when calculating expand).

```ts
const searchedAncestorsKeys: Set<string> = React.useMemo(() => {
        const result: string[] = []
        props.data.forEach(item => {
            bcFilterMatchedAncestors(item as FullHierarchyDataItem, props.data as FullHierarchyDataItem[], textFilters)
                ?.forEach(key => result.push(key))
        })
        return new Set(result)
    }, [props.data, textFilters])
```

```ts
const searchedDescendantsKeys: Set<string> = React.useMemo(() => {
        const result: string[] = []
        props.data.forEach(item => {
            bcFilterMatchedDescendants(item as FullHierarchyDataItem, props.data as FullHierarchyDataItem[], textFilters)
                ?.forEach(key => result.push(key))
        })
        return new Set(result)
    }, [props.data, textFilters])
```
	
	
and got the data like that:
```ts
const filteredData = React.useMemo(() => {
        return textFilters?.length
            ? props.data.filter(item => searchedAncestorsKeys.has(item.id) || searchedDescendantsKeys.has(item.id))
            : props.data
    }, [searchedAncestorsKeys, searchedDescendantsKeys, props.data, textFilters])
```
	
Given the fact of the presence of descendants in the data tree, the proposed algorithm will be changed. It is necessary to take into account that now the required elements can have descendants, but they do not need to be expanded.


Get three subsets of elements and collect them into one:
- manual expanded keys
- expanded keys of associated elements
- filtered expanded keys as all ancestors of filtered elements without last children of subtree

This set of changes leads to expected result and has correctly structured readable code.